chainstate/src/detail/chainstateref.rs

```diff
@@ -422,46 +425,404 @@ impl<'a, S: BlockchainStorageRead, O: OrphanBlocks> ChainstateRef<'a, S, O> {
         Ok(())
     }
 
-    fn check_transactions(&self, block: &Block) -> Result<(), CheckBlockTransactionsError> {
+    fn check_duplicate_inputs(&self, block: &Block) -> Result<(), CheckBlockTransactionsError> {
         // check for duplicate inputs (see CVE-2018-17144)
-        {
-            let mut block_inputs = BTreeSet::new();
-            for tx in block.transactions() {
-                let mut tx_inputs = BTreeSet::new();
-                for input in tx.inputs() {
-                    if !block_inputs.insert(input.outpoint()) {
-                        return Err(CheckBlockTransactionsError::DuplicateInputInBlock(
-                            block.get_id(),
-                        ));
+        let mut block_inputs = BTreeSet::new();
+        for tx in block.transactions() {
+            let mut tx_inputs = BTreeSet::new();
+            for input in tx.inputs() {
+                if !block_inputs.insert(input.outpoint()) {
+                    return Err(CheckBlockTransactionsError::DuplicateInputInBlock(
+                        block.get_id(),
+                    ));
+                }
+                if !tx_inputs.insert(input.outpoint()) {
+                    return Err(CheckBlockTransactionsError::DuplicateInputInTransaction(
+                        tx.get_id(),
+                        block.get_id(),
+                    ));
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn check_duplicate_txs(&self, block: &Block) -> Result<(), CheckBlockTransactionsError> {
+        // check duplicate transactions
+        let mut txs_ids = BTreeSet::new();
+        for tx in block.transactions() {
+            let tx_id = tx.get_id();
+            let already_in_tx_id = txs_ids.get(&tx_id);
+            match already_in_tx_id {
+                Some(_) => {
+                    return Err(CheckBlockTransactionsError::DuplicatedTransactionInBlock(
+                        tx_id,
+                        block.get_id(),
+                    ))
+                }
+                None => txs_ids.insert(tx_id),
+            };
+        }
+        Ok(())
+    }
+
+    fn check_issue_data(
+        &self,
+        token_ticker: &Vec<u8>,
+        amount_to_issue: &Amount,
+        number_of_decimals: &u8,
+        metadata_uri: &Vec<u8>,
+        tx_id: Id<Transaction>,
+        block_id: Id<Block>,
+    ) -> Result<(), CheckBlockTransactionsError> {
+        const MAX_URI_LEN: usize = 1024;
+        const MAX_DEC_COUNT: u8 = 18;
+        const MAX_TICKER_LEN: usize = 5;
```

Comment:
Should these constants be in come config? For example, `ChainstateConfig`.

Reply:
They should be in ChainConfig. I was about to comment about that.

Anything consensus related (meaning, if it changed it'll cause a fork) should go to chainconfig.

Reply:
Moved these to ChainCoing, and I changed names of consts. Please let me know if you will happy with that.